### PR TITLE
Add resources

### PR DESCRIPTION
**Pull request by brramos**
- update acl security rule resource
- add data center resource
- add data center servers resource
- add incap rule resource
- update site resource

### CODE REVIEW

File: incapsula/resource_incap_rule.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Schema: map[string]*schema.Schema{
+			// Required Arguments
+			"enabled": {
+				Description: "Is enabled",
+				Type:        schema.TypeString,
+				Required:    true,
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_incap_rule.go]: will do

File: incapsula/resource_incap_rule.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+			},
+			"name": {
+				Description: "The name",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+
+			// Optional Arguments
+			"site_id": {
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_incap_rule.go]: updating to required

File: incapsula/resource_data_center_servers.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceDataCenterServersDelete(d *schema.ResourceData, m interface{}) error {
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: Delete through api is not working as expected. Following the pattern of setting id to empty string to clear resource when site is deleted. Confirmed data center severs are released when site is deleted.

File: incapsula/resource_data_center.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceDataCenterDelete(d *schema.ResourceData, m interface{}) error {
+	// todo: review delete data center
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: Delete through api is not working as expected. Following the pattern of setting id to empty string to clear resource when site is deleted. Confirmed data center are released when site is deleted.

File: incapsula/resource_incap_rule.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+		},
+
+		Schema: map[string]*schema.Schema{
+			// Required Arguments
+			"enabled": {
+				Description: "Is enabled",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"priority": {
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_incap_rule.go]: 🤔 required? I see in specs that majority have an asterisk for being required but I see a few where it is not required. Not sure to make this optional and with default value of "1"?

File: incapsula/resource_data_center.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: let me look into this more. My example maybe wrong.

File: incapsula/resource_data_center_servers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: let me look into this more. My example maybe wrong.

File: incapsula/resource_incapsula_site.go [hunk 6]
```diff
@@ -62,6 +62,51 @@ func resourceSite() *schema.Resource {
 				Type:        schema.TypeString,
 				Optional:    true,
 			},
+			"site_id": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_incapsula_site.go]: removing

File: incapsula/resource_incapsula_site.go [hunk 7]
```diff
@@ -166,8 +212,112 @@ func resourceSiteRead(d *schema.ResourceData, m interface{}) error {
 }
 
 func resourceSiteUpdate(d *schema.ResourceData, m interface{}) error {
-	// Not implemented
-	return nil
+	client := m.(*Client)
+
+	siteID := d.Get("site_id").(int)
```
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_incapsula_site.go]: updated

File: incapsula/resource_data_center.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center.go]: updated example

File: incapsula/resource_data_center_servers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (brramos) [reply incapsula/resource_data_center_servers.go]: updated example

**Comment by Techcadia** [comment]
@brramos Checking in to see if this might be ready for a merge?

**Comment by trx35479** [comment]
We are looking to use this provider, we need the features from this branch. Just wondering if it is good to merge or will be planning to merge in couple of weeks or so. I will be pulling this branch for now to test our use cases but wondering if there would be a proper release in coming weeks or you will be adding more features?

**Comment by brramos** [comment]
I have addressed all pr comments @Techcadia on April 22. I have no access to test this code anymore.
The last change was making `enabled` an optional value with a default value of true in `resource_incap_rule.go`. My example terraform file was also updated to allow deletion of data center and data center servers.

**Pull request merged**